### PR TITLE
special: Add /dev/full

### PR DESCRIPTION
**Pull request by agkaminski**
## Motivation and Context

Fixes https://github.com/phoenix-rtos/phoenix-rtos-project/issues/197

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: (ia32-generic-qemu).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [x] This PR needs additional PRs to work ([list the PRs, preferably in merge-order](https://github.com/phoenix-rtos/phoenix-rtos-posixsrv/pull/18)).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: special.c [hunk 1]
```diff
@@ -162,6 +205,13 @@ int special_init()
 	err = object_link(o, "/dev/zero");
 	object_put(o);
 
+	if ((o = malloc(sizeof(*o))) == NULL)
```
Comment (badochov) [inline special.c:208]: MISRA

File: special.c [hunk 2]
```diff
@@ -162,6 +205,13 @@ int special_init()
...
+		return -ENOMEM;
+
+	object_create(o, &full_ops);
+	err = object_link(o, "/dev/full");
```
Comment (badochov) [inline special.c:212]: this error value is never checked, same thing in other places in this function

**Review by damianloew — COMMENTED** [review]
I confirm that it solves https://github.com/phoenix-rtos/phoenix-rtos-project/issues/197 and thanks to this change all dd busybox tests pass. IMHO after applying the above suggestions we can merge it.

**Comment by agkaminski** [comment]
Please take a note that https://github.com/phoenix-rtos/phoenix-rtos-posixsrv/pull/18 needs to merged first

File: special.c [hunk 3]
```diff
@@ -160,36 +154,55 @@ static operations_t random_ops = {
...
+	if (err < 0) {
+		return err;
+	}
 
-	if ((o = malloc(sizeof(*o))) == NULL)
-		return -ENOMEM;
+	err = object_link(o, path);
+	if (err < 0) {
+		return err;
+	}
```
Comment (badochov) [inline special.c:200]: Isn't free missing in error paths?

File: special.c [hunk 4]
```diff
@@ -79,12 +79,12 @@ static request_t *zero_read_op(object_t *o, request_t *r)
...
-	int randbuff[16];
-	size_t i, limit, chunk, len = r->msg.o.size;
+	size_t len = r->msg.o.size;
 
 	while (len > 0) {
-		chunk = (len > sizeof(randbuff)) ? sizeof(randbuff) : len;
-		limit = (chunk + sizeof(*randbuff) - 1) / sizeof(*randbuff);
+		int randbuff[16];
+		size_t chunk = (len > sizeof(randbuff)) ? sizeof(randbuff) : len;
+		size_t limit = (chunk + sizeof(*randbuff) - 1) / sizeof(*randbuff);
```
Comment (badochov) [inline special.c:94]: Declarations no longer need to be at the top only in this repo or in all repos?

Comment (agkaminski) [reply special.c:94]: Yup, please take a look at the updated coding guidelines

Comment (badochov) [reply special.c:94]: so happy

Comment (agkaminski) [reply special.c:94]: But excluding kernel, mind you

**Review by badochov — APPROVED** [review]
LGTM

**Review by badochov — APPROVED** [review]
LGTM

**Pull request merged**